#184581179 Two Factor Authentication for Sellers

wilbrord2 opened the pull request:
#### What does this PR do?

This PR verify if user is a seller and he/she can login as a seller

#### Description of Task to be completed?

- [x] A seller will be successfully login with email and password

- [x] A seller should receive email containing random number for 2FA to verify if he/she is owner of stall.

- [x] Document feature using swagger.


#### How should this be manually tested?
- Clone this repo `git clone https://github.com/atlp-rwanda/e-comm-team-amigos-bn`
- Switch your working directory `cd e-comm-team-amigos-bn`
- Create an` .env` file and update it with `.env-example`
- Install all the dependencies by running` npm install`
- Then run the project by typing `npm dev` to start server
- Make sure that postman is installed, open postman or any other API testing tool of your choice
- choose route to be `POST`
- Navigate `localhost:3000/user/login`
- Make sure you created account, enter your username and password
- if you are a seller you will receive email containing `OTP (one-time-password) `
- you can use it to successfully login

#### Endpoints to be tested:

- `POST user/login`

#### Any background context you want to provide?

- PostgreSQL installed in your local machine
- Node JS installed in your local machine

#### What are the relevant pivotal tracker/Trello stories?

Story ID: [#184581179](https://www.pivotaltracker.com/story/show/184581179)

#### Screenshot of  `swagger documentation`  ,  `test`   and  `email format`  to be sent

![Screenshot (134)](https://user-images.githubusercontent.com/79166151/224707737-2312e089-6827-48e7-9858-0525082d4f77.png)

![Screenshot (133)](https://user-images.githubusercontent.com/79166151/224707782-deb469c2-16d5-4ccc-b916-8384da00b692.png)

![email](https://user-images.githubusercontent.com/79166151/224961520-c8598d7f-a430-4d47-8426-ebbb5cf14b4c.png)

wilbrord2 committed:
feat(Authentication):Two Factor Authentication for Sellers
 - verifying if user role is a seller/vendor
 - generating a random OTP for checking and verifying user for second time

 - allowing user to login if he/she is successfully verified
[Delivers #184581179]

nshutijonathan commented on src/tests/auth.test.js:
```diff
@@ -0,0 +1,295 @@
+    // Check that the user's otpcode and otpcodeexpiration have been updated in the database
+    await models.sequelize.sync();
+    const updatedUser = await models.User.findByPk(user.id);
+    // const timestamp = updatedUser.otpcodeexpiration;
```
Don't push commented out codes

nshutijonathan commented on src/tests/auth.test.js:
```diff
@@ -0,0 +1,295 @@
+  //   const transporterSpy = sinon.spy(nodemailer, 'createTransport');
+  //   const mailOptions = {
+  //     from: 'sender@example.com',
+  //     to: user.email,
```
commented out codes should not get pushed

GasanaJr reviewed (changes requested):
The feature works as expected, but the circleci tests must still be included. Integrate the tests, and we will be good to go. Also update the body of your PR to include screenshots for swagger and running tests.

nyakamwe committed:
ft(signUp):sign up a user
- creates signUp endpoint
- validate user inputs
- validate user inputs if are valid
- generate jwt on successfull user registration
- create a unut testing for signUp
- documents user signUp endpoint with swager ui

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

ft(signUp):sign up a user

- creates signUp endpoint
- validate user inputs
- generate jwt on successfull user registration
- creates a middleware function for verifying all protected routes
- documents both user signUp and login

[Delivers #184581172]

ft(login): User Login

- create user login endpoint

[Finishes #184581175]

feat(user): User login and testing

- add user login endpoint
- tests the endpoint using mocha and chai

[Finishes #184581175]

bug(rimraf) resolve rimraf commit error

    -This commit resolves the errors in pachage.json

    [ fixes 184581169 ]

chore(docker compose): setup docker for backend

 -ensure app can run in docker containers
 -create services to run express server and postgres

[Finishes #184581170]

chore(docker compose): setup docker for backend

 -ensure app can run in docker containers
 -create services to run express server and postgres

[Finishes #184581170]

feat(user): User signup

- resolve conflict

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

ft(signUp):sign up a user

- creates signUp endpoint
- validate user inputs
- generate jwt on successfull user registration
- creates a middleware function for verifying all protected routes
- documents both user signUp and login

[Delivers #184581172]

ft(sugn up): create user account
- create  user account endpoint
- add user validation to ensure that user inputs are valid
- unit testing for the user sign up
- document user sign up endpoint with swager ui

[Delivers #1845811]

ft(signUp):sign up a user

- creates signUp endpoint
- validate user inputs
- generate jwt on successfull user registration
- creates a middleware function for verifying all protected routes
- documents both user signUp and login

[Delivers #184581172]

ft(signUp):sign up a user
- creates signUp endpoint
- validate user inputs
- validate user inputs if are valid
- generate jwt on successfull user registration
- create a unut testing for signUp
- documents user signUp endpoint with swager ui

[Delivers #184581172]

ft(signUp):sign up a user
- creates signUp endpoint
- validate user inputs
- validate user inputs if are valid
- generate jwt on successfull user registration
- create a unut testing for signUp
- documents user signUp endpoint with swager ui

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

ft(signUp):sign up a user

- creates signUp endpoint
- validate user inputs
- generate jwt on successfull user registration
- creates a middleware function for verifying all protected routes
- documents both user signUp and login

[Delivers #184581172]

ft(login): User Login

- create user login endpoint

[Finishes #184581175]

feat(user): User login and testing

- add user login endpoint
- tests the endpoint using mocha and chai

[Finishes #184581175]

feat(Verification): Implement send email verification endpoint

- Implement send Email middleware
- Implement verification end point

[Delivers #184581173]

feat(Verification): Implement send email verification endpoint

- Implement send Email middleware
- Implement verification end point

[Delivers #184581173]

 ft(signUp):sign up a user
- creates signUp endpoint
- validate user inputs
- validate user inputs if are valid
- generate jwt on successfull user registration
- create a unut testing for signUp
- documents user signUp endpoint with swager ui

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

ft(signUp):sign up a user

- creates signUp endpoint
- validate user inputs
- generate jwt on successfull user registration
- creates a middleware function for verifying all protected routes
- documents both user signUp and login

[Delivers #184581172]

ft(login): User Login

- create user login endpoint

[Finishes #184581175]

feat(user): User login and testing

- add user login endpoint
- tests the endpoint using mocha and chai

[Finishes #184581175]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

feat(user): User signup

- add user endpoint
- add user validation to ensure that user enters valid email, password and username
- rename index file to user in a folder called controllers

[Delivers #184581172]

ft(signUp):sign up a user

- creates signUp endpoint
- validate user inputs
- generate jwt on successfull user registration
- creates a middleware function for verifying all protected routes
- documents both user signUp and login

[Delivers #184581172]

ft(login): User Login

- create user login endpoint

[Finishes #184581175]

feat(user): User login and testing

- add user login endpoint
- tests the endpoint using mocha and chai

[Finishes #184581175]

GasanaJr reviewed (changes requested):
Solve conflicts you have in develop

AbdulKhaliq59 reviewed (changes requested):
Resolve conflict

nshutijonathan commented on src/database/migrations/20230312135251-Users.js:
```diff
@@ -60,8 +68,20 @@ module.exports = {
         type: Sequelize.DATE
       }
     });
+    // await queryInterface.addColumn('Users', ' otpcode', {
```
remove commented out codes here

nshutijonathan commented on src/database/migrations/20230312135251-Users.js:
```diff
@@ -60,8 +68,20 @@ module.exports = {
+    // });
+
+    // await queryInterface.addColumn('Users', 'otpcodeexpiration', {
+    //   type: Sequelize.STRING,
```
remove commented out codes here

nshutijonathan commented on src/database/migrations/20230312135251-Users.js:
```diff
@@ -60,8 +68,20 @@ module.exports = {
+  async down(queryInterface) {
     await queryInterface.dropTable('Users');
+    // await queryInterface.removeColumn('Users', 'otpcode');
+    // await queryInterface.removeColumn('Users', 'otpcodeexpiration');
```
commented out codes should not be pushed

nshutijonathan reviewed (changes requested):
Remember to rebase so that the conflicts are gone and then we are good to go

nshutijonathan reviewed (approved):
LGTM
